### ISSUE TITLE
Why CancellationToken passed to IHostedService - StartAsync method is not used ?

### ISSUE THREAD

**Issue by shreesharao** [dotnet/extensions#3218]
When you look at `StartAsync` method here -https://github.com/dotnet/extensions/blob/494e2c53cd0ea075ba3783748d62c66bc4a353e2/src/Hosting/Abstractions/src/BackgroundService.cs#L30, you observe that the `CancellationToken` passed is not used when calling `ExecuteAsync` method.

Because IHostedServices are registered as Singleton, this is causing issues, When I am manually trying to start and stop the hosted service as below

```
private readonly IHost _host;
private CancellationTokenSource _cancellationTokenSource;
        
public App()
        {
            _host = Host.CreateDefaultBuilder()
                .ConfigureServices((hostContext, services) =>
                {
                    services.AddHostedService<Worker>()
                }).Build();

            SystemEvents.PowerModeChanged += SystemEvents_PowerModeChanged;
        }
```

```
private async void SystemEvents_PowerModeChanged(object sender, PowerModeChangedEventArgs e)
        {
            var workers = _host.Services.GetServices<IHostedService>();
            Log.Information($"Found IHostedService instances - {workers.ToCSV()}");
            switch (e.Mode)
            {
                case PowerModes.Resume:
                    Log.Warning("System is resuming. Restarting the workers");
                    try
                    {
                        _cancellationTokenSource = new CancellationTokenSource();
                        foreach (var worker in workers)
                        {
                            await worker.StartAsync(_cancellationTokenSource.Token).ConfigureAwait(false);
                        }
                    }
                    catch (Exception ex)
                    {
                        Log.Error(ex, $"{ex.Message}");
                    }
                    break;

                case PowerModes.Suspend:
                    Log.Warning("System is suspending. Stopping the workers");
                    _cancellationTokenSource.Cancel();
                    foreach (var worker in workers)
                    {
                        await worker.StopAsync(_cancellationTokenSource.Token).ConfigureAwait(false);
                    }
                    break;
            }
        }
```

**Worker.cs**
```
public class Worker : BackgroundService
    {
        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
                while (!stoppingToken.IsCancellationRequested)
                {
                    await Task.Delay(TimeSpan.FromSeconds(_appSettings.IntervalInMins), stoppingToken);
                }

                _logger.LogInformation("Cancellation requested. Stopping the activity");
        }
    }

```

When `worker.StartAsync(_cancellationTokenSource.Token)` method is called, I am getting `stoppingToken.IsCancellationRequested` as `true` because the token I am passing to StartAsync method is not used.

How do I achieve this behavior?

**Comment by klettier** [comment]
I already did a [Pr](https://github.com/dotnet/extensions/pull/2823) and the propagation of the cancellation token will be in the next release.

In the meantime I would recommend to inherit from **IHostedService** instead of **BackGroundService**

**Comment by shreesharao** [comment]
@klettier , Thank you for pointing me out to the PR. I am adding `BackgroundService.cs` directly to my application. Taken the code from here -  https://github.com/dotnet/runtime/blob/6ef538d46fc971f351a22cbe854717f4ada59d89/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs 

You think doing it this way would cause any issue?

**Comment by klettier** [comment]
No I think it's fine that way.

Concerning the code you showed I would remove the call to ```_host.Services.GetServices<IHostedService>()``` inside the event callback like so:

```csharp
static void App()
{
    var host = Host.CreateDefaultBuilder()
        .ConfigureServices((hostContext, services) =>
        {
            services.AddHostedService<Worker>();
        }).Build();

    var workers = host.Services.GetServices<IHostedService>().ToArray();

    SystemEvents.PowerModeChanged += (sender, e) => SystemEvents_PowerModeChanged(workers, sender, e);
}

private static async void SystemEvents_PowerModeChanged(IHostedService[] workers, object sender, PowerModeChangedEventArgs e)
{
    Log.Information($"Found IHostedService instances - {workers.ToCSV()}");
    switch (e.Mode)
    {
        case PowerModes.Resume:
            Log.Warning("System is resuming. Restarting the workers");
            try
            {
                _cancellationTokenSource = new CancellationTokenSource();
                foreach (var worker in workers)
                {
                    await worker.StartAsync(_cancellationTokenSource.Token).ConfigureAwait(false);
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex, $"{ex.Message}");
            }
            break;

        case PowerModes.Suspend:
            Log.Warning("System is suspending. Stopping the workers");
            _cancellationTokenSource.Cancel();
            foreach (var worker in workers)
            {
                await worker.StopAsync(_cancellationTokenSource.Token).ConfigureAwait(false);
            }
            break;
    }
}
```

**Comment by shreesharao** [comment]
@klettier , Thank you, I will do the change you suggested. But what is the problem getting service instances inside event callback. Is it a problem doing that?

**Comment by davidfowl** [comment]
User code shouldn’t be calling start and stop on IHostedService instances

**Comment by shreesharao** [comment]
@davidfowl , then how to handle the case where one need to stop and restart one of the long running worker services? like in my case, I want to perform some activity when the system resumes.

**Comment by davidfowl** [comment]
That’s not a supported feature. I’d use another type. There’s no benefit to using hosted services if you’re the one calling start and stop, that’s what the host is supposed to do.

**Comment by shreesharao** [comment]
@davidfowl , yes I understand your point. Even in my application, `host` is the one which handles the lifetime of all hosted services. But _based on user input or any other external event, stopping/starting one of the many hosted services is also an use case I think_. And it is implicitly supported by the way worker services designed.

**Comment by anurse** [comment]
Closing this as the original request was resolved by #2823.
